distributions/openhab/src/main/resources/bin/update.ps1

```diff
@@ -69,190 +12,582 @@ Function Update-openHAB {
     The version to upgrade to.
     .PARAMETER Snapshot
     Upgrade to a snapshot version ($true) or a release version ($false) (default: $false)
+    .PARAMETER AutoConfirm
+    Automatically confirm update (used for headless mode)
     .PARAMETER SkipNew
     Internal use only. For skipping the check for a new update.ps1
     .EXAMPLE
     Update the openHAB distribution in the current directory to the current stable version
     Update-openHAB
     .EXAMPLE
     Update the openHAB distribution in the C:\oh-snapshot directory to the next snapshot version
-    Update-openHAB -OHDirectory C:\oh-snapshot -OHVersion 2.2.0 -Snapshot $true
-    #>
+    Update-openHAB -OHDirectory C:\oh-snapshot -OHVersion 2.3.0 -Snapshot $true
+#>
 
+Function Update-openHAB() {
     [CmdletBinding()]
     param(
-        [Parameter(ValueFromPipeline=$True)]
+        [Parameter(ValueFromPipeline = $True)]
         [string]$OHDirectory = ".",
-        [Parameter(ValueFromPipeline=$True)]
-        [string]$OHVersion = "2.1.0",
-        [Parameter(ValueFromPipeline=$True)]
+        [Parameter(ValueFromPipeline = $True)]
+        [string]$OHVersion,
+        [Parameter(ValueFromPipeline = $True)]
         [boolean]$Snapshot = $false,
-        [Parameter(ValueFromPipeline=$True)]
-        [boolean]$SkipNew = $false
+        [Parameter(ValueFromPipeline = $True)]
+        [boolean]$AutoConfirm = $false,
+        [Parameter(ValueFromPipeline = $True)]
+        [boolean]$SkipNew = $false,
+        [Parameter(ValueFromPipeline = $True)]
+        [boolean]$KeepUpdateScript = $false  # sssh - secret switch ;)
     )
 
-    begin {}
-    process {
+    function DownloadFiles() {
+        param(    
+            [Parameter(Mandatory = $True)]
+            [string] $DownloadSource, 
+            [Parameter(Mandatory = $True)]
+            [string] $OutputFile
+        ) 
+    
+        $uri = New-Object "System.Uri" "$DownloadSource"
+
+        $request = [System.Net.HttpWebRequest]::Create($uri)
+        $request.set_Timeout(15000)
 
-        if (!([Security.Principal.WindowsPrincipal] [Security.Principal.WindowsIdentity]::GetCurrent()
-            ).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
-            throw "This script must be run as an Administrator. Start PowerShell with the Run as Administrator option"
+        Invoke-WebRequest $uri -Outfile $Outputfile -ErrorAction Stop
+    }
+
+    function NormalizeVersionNumber() {
```

Comment:
There's this cool new function but it's not actually used before we try to download files that don't exist... Like `upgrade.bat 2.3`

Reply:
Also, going from 2.3.0 to 2.3.0 would be a great way to do a reinstall but we should do some validation/notification that there isn't really an 'upgrade' happening.

No error handling when file just doesn't exist. Like Stable 2.4.0.

Reply:
I like it (ie reinstalling).  But what do you think we should do with AutoConfirm=true if we are reinstalling?  

1. Do we assume the users intention was to reinstall and just give a warning message about reinstallation (rather than a confirmation since it's autoconfirm)?
2. Or do we consider this situation special enough that we ignore the autoconfirm on a reinstallation and ask the user?

Reply:
Ugh... Well, if we just go ahead and do it, what's the worst thing that happens? They stay where they were. I say just go for it.